feat: validate input and fhir response

d-rita opened the pull request:
Implements [DEVREL-71](https://dhis2.atlassian.net/browse/DEVREL-71)

**Description**
This PR does a couple of things in terms of validation and sanitisation:
- It validates the structure of the  jsonata expression returned from evaluating the fhir response to ensure it is not empty before setting the other fields in the form. 
- It sanitises the values mapped into the jsonata expression. Done by trimming and removing html tags if any. This is a crucial step against all kinds of data that may be returned from the external server being queried. 
- It validates the ID value used to query the civil registry. Ensures it is non-empty at the very least.

**Tests**
- Added tests for invalid ID inputs that require validation, an invalid fhir response that is caught early on, and "unexpected" values with white spaces or html tags,  requiring sanitisation before being filled into the form. 

[DEVREL-71]: https://dhis2.atlassian.net/browse/DEVREL-71?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

d-rita committed:
feat: validate input and fhir response

d-rita committed:
chore: update failing test

d-rita committed:
feat: use dompurify to sanitise mapped fhir params

d-rita committed:
chore: add tests for validation and sanitisation

d-rita committed:
Merge branch 'main' into DEVREL-71/security-and-validation

kabaros commented on civil-registry-plugin/src/LookupField/LookupField.tsx:
```diff
@@ -96,17 +113,41 @@ export const LookupField = ({
     }, [personMapData])
 
     const handleSearch = useCallback(async () => {
+        if (!isValidValue(patientId)) {
```
I think the validation should happen after the sanitisation? Otherwise, you could pass here, but then sanitise and end up with an empty string

d-rita replied:
Rectified this. Sanitised the id first then validation happens in this step.

kabaros commented on civil-registry-plugin/src/LookupField/LookupField.tsx:
```diff
@@ -31,6 +32,20 @@ const registryErrMessage = i18n.t(
+        return false
+    }
+
+    const alphanumericPattern = /^[a-zA-Z0-9]+$/i
```
following up on the Slack discussion, I think we should relax this pattern as we're not sure of the format (maybe it contains dots or @ sign, slashes in personal IDs and passport numbers are possible...)

d-rita replied:
Updated the pattern to allow dots, @ signs, and slashes on top of numbers and letters. Is this sufficient or do we not need the pattern as much?

kabaros commented on civil-registry-plugin/src/LookupField/LookupField.tsx:
```diff
@@ -1,5 +1,6 @@
 import i18n from '@dhis2/d2-i18n'
 import { Button, Help, Input, Label, Tooltip } from '@dhis2/ui'
+import DOMPurify from 'dompurify';
```
we could consider https://developer.mozilla.org/en-US/docs/Web/API/HTML_Sanitizer_API if we're OK with limited browser support (and avoid the extra library)

(not necessary to tackle now)

KaiVandivier commented on civil-registry-plugin/src/LookupField/LookupField.tsx:
```diff
@@ -96,17 +113,41 @@ export const LookupField = ({
+            console.error("Invalid data returned from registry lookup")
+            setMappingError(true)
+            return
+        }
```
This will be covered by the `jsonata` actions in the next lines -- no need to validate here

d-rita replied:
Removed the validation entirely. Letting jsonata actions handle it.

d-rita committed:
fix: improve sanitisation and validation logic

- expand range of allowed characters for the ID
- remove redundant jsonata validation check
- sanitise ID before validation

d-rita committed:
chore: update invalid ID test

d-rita committed:
feat: use generic ID validation

- only ensure the ID field is populated
- remove invalid characters' unit test

d-rita committed:
chore: update failing tests

- validate against whitespace ID input for query to fail
- validate against empty jsonata expression from fhir response

d-rita committed:
chore: update README with ID validation security note
